src/tools/pkcs11-tool.c

```diff
@@ -1870,6 +1889,89 @@ static void decrypt_data(CK_SLOT_ID slot, CK_SESSION_HANDLE session,
 		util_fatal("Cannot read from %s: %m", opt_input);
 	in_len = r;
 
+	/* set "default" MGF and hash algorithms. We can overwrite MGF later */
+	switch (opt_mechanism) {
+	case CKM_RSA_PKCS_OAEP:
+		oaep_params.hashAlg = opt_hash_alg;
+		switch (opt_hash_alg) {
+		case CKM_SHA224:
+			oaep_params.mgf = CKG_MGF1_SHA224;
+			break;
+		case CKM_SHA256:
+			oaep_params.mgf = CKG_MGF1_SHA256;
+			break;
+		case CKM_SHA384:
+			oaep_params.mgf = CKG_MGF1_SHA384;
+			break;
+		case CKM_SHA512:
+			oaep_params.mgf = CKG_MGF1_SHA512;
+			break;
+		default: 
+			oaep_params.hashAlg = CKM_SHA_1;
+			/* fallthrough to SHA-1 default */
+		case CKM_SHA_1:
+			oaep_params.mgf = CKG_MGF1_SHA1;
+			break;
+		}
+		break;
+
+#if 0 /* we do not have these definitions yet! */
+	case CKM_SHA1_RSA_PKCS_OAEP:
+		oaep_params.hashAlg = CKM_SHA_1;
+		oaep_params.mgf = CKG_MGF1_SHA1;
+		break;
+
+	case CKM_SHA224_RSA_PKCS_OAEP:
+		oaep_params.hashAlg = CKM_SHA224;
+		oaep_params.mgf = CKG_MGF1_SHA224;
+		break;
+
+	case CKM_SHA256_RSA_PKCS_OAEP:
+		oaep_params.hashAlg = CKM_SHA256;
+		oaep_params.mgf = CKG_MGF1_SHA256;
+		break;
+
+	case CKM_SHA384_RSA_PKCS_OAEP:
+		oaep_params.hashAlg = CKM_SHA384;
+		oaep_params.mgf = CKG_MGF1_SHA384;
+		break;
+
+	case CKM_SHA512_RSA_PKCS_OAEP:
+		oaep_params.hashAlg = CKM_SHA512;
+		oaep_params.mgf = CKG_MGF1_SHA512;
+		break;
+#endif
+	case CKM_RSA_PKCS:
+		mech.pParameter = NULL;
+		mech.ulParameterLen = 0;
+		break;
+	default:
+		util_fatal("Illegal mechanism %s for RSA-OAEP\n", p11_mechanism_to_name(opt_mechanism));
+	}
+
+
+	/* One of RSA-OAEP mechanisms above: They need parameters */
```

Comment:
There is only one RSA-OAEP at this moment.

Reply:
Yeah...? ;-)